optimize/protocol.py

```diff
@@ -0,0 +1,254 @@
+        delta_us = vec_delta_u(prev_lam, next_lam)
+
+        stddev_estimate = reweighted_stddev(
+            f_n=delta_us,
+            target_logpdf_n=target_logpdf_n,
+            source_logpdf_n=reference_log_weights_n,
+        )
+
+        return stddev_estimate
+
+    return work_stddev_estimator
+
+
+def reweighted_stddev(f_n: Array, target_logpdf_n: Array, source_logpdf_n: Array) -> Float:
+    """Compute reweighted estimate of
+    stddev(f(x)) under x ~ p_target
+    based on samples   x ~ p_source
+
+    where
+        p_target(x) = exp(target_logpdf(x)) / Z_target
+
+    using samples from a different source
+        x_n ~ p_source
+        where
+        p_source(x) = exp(source_logpdf(x)) / Z_source
+
+    The inputs are arrays "{fxn_name}_n" containing the result of
+    calling each fxn on a fixed array of samples:
+
+    * f_n = [f(x_n) for x_n in samples]
+    * target_logpdf_n = [target_logpdf(x_n) for x_n in samples]
+    * source_logpdf_n = [source_logpdf(x_n) for x_n in samples]
+    """
+
+    log_weights_n = target_logpdf_n - source_logpdf_n
+    weights = np.exp(log_weights_n - logsumexp(log_weights_n)).flatten()
+
+    f_mean = np.sum(weights * f_n)
+    squared_deviations = (f_n - f_mean) ** 2
+
+    # sanitize 0 * inf -> 0 (instead of nan)
+    weighted_squared_deviations = weights * squared_deviations
+    sanitized = np.nan_to_num(weighted_squared_deviations, nan=0)
+    stddev = np.sqrt(np.sum(sanitized))
+
+    return stddev
+
+
+def construct_heuristic_lambda_pair_assessor(work_stddev_estimator) -> StepAssessor:
+    """Construct a function f(prev_lam, trial_next_lam) where bisection search on second argument
+    can be used to select next_lam so that p(x|next_lam) is a specified "distance" from p(x|prev_lam)"""
+
+    def assess_lambda_pair(prev_lam, next_lam, desired_stddev=1.0, max_step=0.25):
+        """if (next_lam - prev_lam <= max_step), compute (max(forward_stddev, reverse_stddev) - desired_stddev)"""
+        too_far = next_lam - prev_lam > max_step
+        if too_far:
+            return +np.inf
+
+        # compute max of forward, reverse work stddevs
+        forward_stddev = work_stddev_estimator(prev_lam, next_lam)
+        reverse_stddev = work_stddev_estimator(next_lam, prev_lam)
+        higher_stddev = max(forward_stddev, reverse_stddev)
+
+        return higher_stddev - desired_stddev
+
+    return assess_lambda_pair
+
+
+def greedily_optimize_protocol(assess_lambda_pair: StepAssessor, max_iterations=1000) -> Array:
+    """Optimize a lambda protocol from "left to right"
+
+    Sequentially pick next_lam so that
+    assess_lambda_pair(prev_lam, next_lam) ~= 0
+
+    assess_lambda_pair(prev_lam, next_lam) might compute
+    some measure of distance between p(x | prev_lam) and p(x | next_lam)
+    and compare that distance to a desired spacing `distance - target_distance`
+        (returning a negative number if too close, a positive number if too far)
```

Comment:
And returning 0 if just right.

A more descriptive name than "step_assessor"/"assess_lambda_pair"/... would hint more that the function is destined for [`scipy.optimize.bisect`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.bisect.html)ion. (The function must change sign with increasing step size, and the location of the sign change should encode the desired step size.)

Reply:
I can't think of anything I'm very happy with either, although it seems like there should be a standard name for this! Maybe `comparator`?

Reply:
Alternatively, I wonder if you could avoid needing to name this by pulling the subtraction out of `assess_lambda_pair` (renaming to `thermodyamic_distance` or similar) and then changing the "comparator" for bisection to `thermodynamic_distance(prev_lam, trial_lam) - target_distance`?

Reply:
I like this alternative (and I think specifying the `target_distance` separately would make it closer to [the signature](https://github.com/choderalab/yank/blob/59fc6313b3b7d82966afc539604c36f4db9b952c/Yank/pipeline.py#L2284) of the thermodynamic trailblazing implementation in yank).

I was getting hung up on wanting to allow new distance measures (for example, defining `distance(prev_lam, trial_lam)` to be a cheap reweighted estimate of the BAR uncertainty, rather than a cheap reweighted estimate of the work stddev), and to allow the `target_distance` to be a function of `prev_lam` rather than a constant.

Reply:
Addressed in https://github.com/proteneer/timemachine/pull/538/commits/e284195ed51184e7be11fb09c88b107cbc8856ca

Reply:
A good sign: applying this change allows me to [delete](https://github.com/proteneer/timemachine/pull/538/commits/2283ea290ba6e7f9c892cf94b366ebd2791ce71a) some wordy comments